flow/layers/container_layer.cc

```diff
@@ -6,13 +6,42 @@
 
 namespace flutter {
 
-ContainerLayer::ContainerLayer() {}
+ContainerLayer::ContainerLayer()
+    : child_needs_screen_readback_(false), renders_to_save_layer_(false) {}
 
 ContainerLayer::~ContainerLayer() = default;
 
 void ContainerLayer::Add(std::shared_ptr<Layer> layer) {
-  layer->set_parent(this);
+  Layer* the_layer = layer.get();
+  the_layer->set_parent(this);
   layers_.push_back(std::move(layer));
+  if (the_layer->tree_reads_surface()) {
+    UpdateChildReadback(the_layer);
+  }
+}
+
+void ContainerLayer::UpdateChildReadback(Layer* layer) {
+  if (child_needs_screen_readback_ == layer->tree_reads_surface()) {
+    return;
+  }
+  if (child_needs_screen_readback_) {
+    // Transitioning to false, but only if there are no other children
```

Comment:
I believe the complexity here is caused by `ClearChildren` where we allow `Layer` to mutate its children. Eventually, we'd like to make `Layer` completely immutable so this logic could be saved (https://github.com/flutter/flutter/issues/33807).

For now, maybe we can change `child_needs_screen_readback_` to something like `child_maybe_need_screen_readback_` so it can only turn to `true` from `false`, but not the other way around.

It's Ok to sometimes return true but it's actually false. I think with our current use cases of `ClearChildren` (which just rewrap all children in another `ContainerLayer`), we would not actually get a layer tree that returns true when it's actually false.

Reply:
Making the children immutable would complicate the work I'm doing to track dirty regions - I rely on replacing children to determine if the children are the same from frame to frame. I can see if that can be done with immutable children, but right now that's how I approached it.

Reply:
If we can rearrange the frame pumping to allow us to choose a destination after PreRoll (outside the scope of this current change) then we could move all of this logic into the PreRoll context and it would no longer require any mutability in the layers themselves, nor any propagation of flags into parents.

Reply:
If we unravel the PreRoll/AcquireFrame ordering in the frame pumping then these methods and comments will become obsolete. There doesn't seem to be movement to make layers immutable at this time, so I believe that the changes which make this obsolete may happen before that becomes an issue and if the immutable layers happen sooner, it would be a simple change to just make the flags be conservatively true at that time.

For now, the implementation matches the current state of the layers not really being immutable.

Reply:
Sure, matching the current state sounds good to me.

My other concern is about the performance --- the asymptotic complexity of modifying N children in a container layer.

I understand that in most common cases, the complexity is linear (e.g., just call `Add` N times).

However, consider the following case
1. Initially, there's a `ContainerLayer` with N `ContainerLayer` children, each of which needs readback (e.g., each child has a `BackdropFilter` grand child).
2. We call `ClearChildren` for each child `ContainerLayer`, and thus make that child no longer need readback, and call the parent `ContainerLayer`'s `UpdateChildReadback`.
3. In the i-th `UpdateChildReadback` call, we has to go over i-1 older siblings in the for loop below to eventually figure out that the parent `ContainerLayer` still needs readback.

The overall complexity in that case seems to be O(N^2), which is something that Flutter tries very hard to avoid.

That case may not be possible in the current API that we expose to the Dart world. However, I think we still need to be very careful about that. Can you please confirm if the O(N^2) is indeed the case above? If so, maybe let's have an offline discussion about how to handle that.

Reply:
It is true that the potential complexity is O(N^2) and I agree that is a bad thing in general. In this particular case, N should be quite low since BackdropFilters are very rare. I could revert to the counter logic that I used to have which I felt might create a potential to become out-of-date as the source code changes, but I think there is a simpler solution:

Currently no layers (and thus, no "child layers") in the system have any way to go from "readback=true" to "readback=false". That sequence is impossible in the current code base. While there may be some unknown use case in the future that makes that transition, for the immediate future the only realistic way for such a transition to happen would be for BDF to implement caching and use that state to determine its readback potential per frame. In order for that to happen, it would need to be able to report its findings from its PreRoll method - but this mechanism doesn't support modification during PreRoll because the decision about surface types is already made before PreRoll. In addition, we would also have to implement a layer stability mechanism as is being worked on to determine update regions in order for BDF to even know if it can cancel its readback condition. That is 2 significant changes to the current frame pumping logic before we can get to a point where a child changing from true to false becomes possible. (And I don't think we have any other mechanisms currently planned or on the horizon that would exhibit that behavior either.)

This transition case is therefore predicated upon (at least) moving surface decisions until after PreRoll. At that point, we can completely replace this mechanism with a much better one that works during PreRoll and stores all information in the PreRoll Context. That new mechanism would avoid these flags in the Layers themselves and would not need to propagate any information to parents. All in all, that would be a simpler and faster mechanism in general and it would have no similar quadratic behaviors in it.

Given all of that, I think the best approach right now would be to eliminate the "child true->false" path entirely and, as you said, only allow the property to change from false to true during Add() or during propagation. It would still reset to false during ClearChildren() though, but there is no coordination between children required there, it is a decisive change to false.

I'll update the code to eliminate worrying about children changing from readback to non-readback in a pending commit.

Reply:
I've changed the child tracking to only accept transitions to true (except in ClearChildren). I also updated the name of the method to NotifyChildReadback().